### ISSUE TITLE
`master` build fails with `-std=c++17` Apple gcc flag

### ISSUE THREAD

**Issue by ikitayama** [nest/nest-simulator#2252]
I need to use a polymorphic allocator supported in C++17, but NEST build fails with this:

If the sli subsystem expert can take a look at this, that'd be great. Thanks!

```
[ 32%] Building CXX object nestkernel/CMakeFiles/nestkernel.dir/nest_names.cpp.o
[ 32%] Building CXX object nestkernel/CMakeFiles/nestkernel.dir/nestmodule.cpp.o
In file included from /Users/itaru/projects/nest-simulator/nestkernel/nestmodule.cpp:23:
In file included from /Users/itaru/projects/nest-simulator/nestkernel/nestmodule.h:27:
In file included from /Users/itaru/projects/nest-simulator/nestkernel/event.h:29:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/algorithm:653:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/memory:679:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1407:52: error: implicit instantiation of undefined template 'std::tuple_size<lockPTRDatum<Dictionary, &SLIInterpreter::Dictionarytype>>'
_LIBCPP_INLINE_VAR constexpr size_t tuple_size_v = tuple_size<_Tp>::value;
                                                   ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1427:39: note: in instantiation of variable template specialization 'std::tuple_size_v<lockPTRDatum<Dictionary, &SLIInterpreter::Dictionarytype>>' requested here
        typename __make_tuple_indices<tuple_size_v<remove_reference_t<_Tuple>>>::type{})
                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1423:26: note: in instantiation of exception specification for 'apply<sharedPtrDatum<nest::Parameter, &nest::NestModule::ParameterType> &, lockPTRDatum<Dictionary, &SLIInterpreter::Dictionarytype> &>' requested here
constexpr decltype(auto) apply(_Fn && __f, _Tuple && __t)
                         ^
/Users/itaru/projects/nest-simulator/nestkernel/nestmodule.cpp:1999:17: note: in instantiation of function template specialization 'std::apply<sharedPtrDatum<nest::Parameter, &nest::NestModule::ParameterType> &, lockPTRDatum<Dictionary, &SLIInterpreter::Dictionarytype> &>' requested here
  auto result = apply( param, positions );
                ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/__tuple:24:50: note: template is declared here
template <class _Tp> struct _LIBCPP_TEMPLATE_VIS tuple_size;
                                                 ^
In file included from /Users/itaru/projects/nest-simulator/nestkernel/nestmodule.cpp:23:
In file included from /Users/itaru/projects/nest-simulator/nestkernel/nestmodule.h:27:
In file included from /Users/itaru/projects/nest-simulator/nestkernel/event.h:29:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/algorithm:653:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/memory:679:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1427:39: error: non-type template argument is not a constant expression
        typename __make_tuple_indices<tuple_size_v<remove_reference_t<_Tuple>>>::type{})
                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1423:26: note: in instantiation of exception specification for 'apply<sharedPtrDatum<nest::Parameter, &nest::NestModule::ParameterType> &, lockPTRDatum<Dictionary, &SLIInterpreter::Dictionarytype> &>' requested here
constexpr decltype(auto) apply(_Fn && __f, _Tuple && __t)
                         ^
/Users/itaru/projects/nest-simulator/nestkernel/nestmodule.cpp:1999:17: note: in instantiation of function template specialization 'std::apply<sharedPtrDatum<nest::Parameter, &nest::NestModule::ParameterType> &, lockPTRDatum<Dictionary, &SLIInterpreter::Dictionarytype> &>' requested here
  auto result = apply( param, positions );
                ^
In file included from /Users/itaru/projects/nest-simulator/nestkernel/nestmodule.cpp:23:
In file included from /Users/itaru/projects/nest-simulator/nestkernel/nestmodule.h:27:
In file included from /Users/itaru/projects/nest-simulator/nestkernel/event.h:29:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/algorithm:653:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/memory:679:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1407:52: error: implicit instantiation of undefined template 'std::tuple_size<sharedPtrDatum<nest::NodeCollection, &nest::NestModule::NodeCollectionType>>'
_LIBCPP_INLINE_VAR constexpr size_t tuple_size_v = tuple_size<_Tp>::value;
                                                   ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1427:39: note: in instantiation of variable template specialization 'std::tuple_size_v<sharedPtrDatum<nest::NodeCollection, &nest::NestModule::NodeCollectionType>>' requested here
        typename __make_tuple_indices<tuple_size_v<remove_reference_t<_Tuple>>>::type{})
                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1423:26: note: in instantiation of exception specification for 'apply<sharedPtrDatum<nest::Parameter, &nest::NestModule::ParameterType> &, sharedPtrDatum<nest::NodeCollection, &nest::NestModule::NodeCollectionType> &>' requested here
constexpr decltype(auto) apply(_Fn && __f, _Tuple && __t)
                         ^
/Users/itaru/projects/nest-simulator/nestkernel/nestmodule.cpp:2014:17: note: in instantiation of function template specialization 'std::apply<sharedPtrDatum<nest::Parameter, &nest::NestModule::ParameterType> &, sharedPtrDatum<nest::NodeCollection, &nest::NestModule::NodeCollectionType> &>' requested here
  auto result = apply( param, nc );
                ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/__tuple:24:50: note: template is declared here
template <class _Tp> struct _LIBCPP_TEMPLATE_VIS tuple_size;
                                                 ^
In file included from /Users/itaru/projects/nest-simulator/nestkernel/nestmodule.cpp:23:
In file included from /Users/itaru/projects/nest-simulator/nestkernel/nestmodule.h:27:
In file included from /Users/itaru/projects/nest-simulator/nestkernel/event.h:29:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/algorithm:653:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/memory:679:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1427:39: error: non-type template argument is not a constant expression
        typename __make_tuple_indices<tuple_size_v<remove_reference_t<_Tuple>>>::type{})
                                      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/tuple:1423:26: note: in instantiation of exception specification for 'apply<sharedPtrDatum<nest::Parameter, &nest::NestModule::ParameterType> &, sharedPtrDatum<nest::NodeCollection, &nest::NestModule::NodeCollectionType> &>' requested here
constexpr decltype(auto) apply(_Fn && __f, _Tuple && __t)
                         ^
/Users/itaru/projects/nest-simulator/nestkernel/nestmodule.cpp:2014:17: note: in instantiation of function template specialization 'std::apply<sharedPtrDatum<nest::Parameter, &nest::NestModule::ParameterType> &, sharedPtrDatum<nest::NodeCollection, &nest::NestModule::NodeCollectionType> &>' requested here
  auto result = apply( param, nc );
                ^
4 errors generated.
make[2]: *** [nestkernel/CMakeFiles/nestkernel.dir/nestmodule.cpp.o] Error 1
make[1]: *** [nestkernel/CMakeFiles/nestkernel.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by ikitayama** [comment]
```
❯ gcc -v
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 13.0.0 (clang-1300.0.29.30)
Target: arm64-apple-darwin21.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by heplesser** [comment]
I can reproduce this with Clang 13.0.0 and GCC 11.2.0 when switching to C++17 (requires at the moment, as far as I can see, to manually edit [this line](https://github.com/nest/nest-simulator/blob/06f9276ecef401b0e5cf2cfef8a6f028470ba1c1/cmake/DefaultCompilerFlags.cmake#L25).

@hakonsbm Could you take a look?

**Comment by hakonsbm** [comment]
@heplesser Looks like the compiler confuses `apply()` in nest.h with `std::apply()` from `<tuple>`, but I can't see where we specify the `std` namespace. Changing the calls in nestmodule.cpp to `nest::apply()` seems to fix the problem (at least with GCC 10.3.0 and Clang 12.0.0), but a cleaner solution would be to limit the namespace.

**Comment by heplesser** [comment]
@hakonsbm Thanks for the detective work. A quick search through the NEST codebase did not turn up any worrisome `using namespace std;` line, so I find this surprising. Could there be a "namespace leak" from the compiler, i.e., that `std::apply()` is available in the global namespace, causing the confusion? Given that GCC and Clang have different STL implementations, that would be a bit surprising, but one never knows. For now, I'd suggest to fix this by adding the `nest::` qualifier and a comment explaining it. If you have some spare time, you could try to create a minimal reproducer to see if it is really a problem coming from the STL.
